Board Editor Layout

SJuliez opened the pull request:
Makes the BoardEditor adapt to the gui scaling. Also some code cleanup.
Also moves the reaction to gui scale changes (writing them to the GUI Prefs and creating a Prefchange event) via menu or keyboard shortcut from the ClientGUI (which only operates in lobby and game) to the commonmenubar itself which operates everywhere.

Windchild292 commented on megamek/src/megamek/client/ui/swing/CommonMenuBar.java:
```diff
@@ -826,6 +826,20 @@ private JMenuItem createMenuItem(JMenu m, String label, String command) {
     public void actionPerformed(ActionEvent event) {
+        
+        // Changes that are independent of the current state of MM
+        if (event.getActionCommand().equals(ClientGUI.VIEW_INCGUISCALE)) {
+            float guiScale = GUIPreferences.getInstance().getGUIScale();
+            if (guiScale < ClientGUI.MAX_GUISCALE) {
+                GUIPreferences.getInstance().setValue(GUIPreferences.GUI_SCALE, guiScale + 0.1);
+            }
+        } else if (event.getActionCommand().equals(ClientGUI.VIEW_DECGUISCALE)) {
+            float guiScale = GUIPreferences.getInstance().getGUIScale();
+            if (guiScale > ClientGUI.MIN_GUISCALE) {
+                GUIPreferences.getInstance().setValue(GUIPreferences.GUI_SCALE, guiScale - 0.1);
+            }
+        }
```
Should we return at the end of these two, even if the if statement doesn't occur?

SJuliez replied:
You confuse me :) I thought about returning when the action command is in fact one of the GUISCALE things. But I thought well why not pass it on, if for whatever reason the other components want to do something with it as well. So, no return there, the event is always passed on to e.g. the ClientGUI, even if its a GUISCALE event.

Windchild292 replied:
I missed the word internal, my apologies... but that makes a lot of sense :)

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -299,9 +279,19 @@ public void setTerrainTypes(List<TerrainTypeHelper> terrainTypes) {
+    private FixedYPanel panelBoardSettings = new FixedYPanel();
+    
+    // Help Texts
+    JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"),SwingConstants.LEFT);
+    JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"),SwingConstants.LEFT);
```
```suggestion
    private JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"), SwingConstants.LEFT);
    private JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"), SwingConstants.LEFT);
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -544,37 +530,36 @@ public void windowClosing(WindowEvent e) {
-    private JButton prepareButton(String iconName, String buttonName, ArrayList<JButton> bList) {
-        JButton button = new JButton(buttonName);
-        button.addActionListener(this);
+    private ScalingIconButton prepareButton(String iconName, String buttonName, 
+            ArrayList<ScalingIconButton> bList, int width) {
```
```suggestion
    private @Nullable ScalingIconButton prepareButton(String iconName, String buttonName, 
            List<ScalingIconButton> bList, int width) {
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -584,92 +569,84 @@ private JButton prepareButton(String iconName, String buttonName, ArrayList<JBut
-        JToggleButton button = new JToggleButton(buttonName);
-        button.addActionListener(this);
-        
+    private ScalingIconToggleButton prepareToggleButton(String iconName, String buttonName, 
+            ArrayList<ScalingIconToggleButton> bList, int width) {
```
```suggestion
    private @Nullable ScalingIconToggleButton prepareToggleButton(String iconName, String buttonName, 
            List<ScalingIconToggleButton> bList, int width) {
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -2164,6 +2032,58 @@ public void valueChanged(ListSelectionEvent event) {
+        labHelp2.setFont(scaledFont);
+        labTheme.setFont(scaledFont);
+        
+        ((TitledBorder)panelBoardSettings.getBorder()).setTitleFont(scaledFont);
+        ((TitledBorder)panelHexSettings.getBorder()).setTitleFont(scaledFont);
+        ((TitledBorder)panelTerrSettings.getBorder()).setTitleFont(scaledFont);
```
```suggestion
        ((TitledBorder) panelBoardSettings.getBorder()).setTitleFont(scaledFont);
        ((TitledBorder) panelHexSettings.getBorder()).setTitleFont(scaledFont);
        ((TitledBorder) panelTerrSettings.getBorder()).setTitleFont(scaledFont);
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -2378,4 +2300,95 @@ public int getNumber() {
+            }
+        }
+        
+        void setRolloverImage(Image image) {
+            baseRolloverImage = image;
+        }
+        
+        void setDisabledImage(Image image) {
+            baseDisabledImage = image;
```
```suggestion
        void setRolloverImage(@Nullable Image image) {
            baseRolloverImage = image;
        }
        
        void setDisabledImage(@Nullable Image image) {
            baseDisabledImage = image;
```

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -2378,4 +2300,95 @@ public int getNumber() {
+            }
+        }
+        
+        void setRolloverImage(Image image) {
+            baseRolloverImage = image;
+        }
+        
+        void setSelectedImage(Image image) {
```
```suggestion
        void setRolloverImage(@Nullable Image image) {
            baseRolloverImage = image;
        }
        
        void setSelectedImage(@Nullable Image image) {
```

Windchild292 reviewed (commented):
Looks good, just a few comments first. Also, can we try using @Nullable whenever there is a potential null return, as that should help prevent some of the NPEs we get in the future.

SJuliez commented:
@Windchild292 changes are in; I changed a few things concerning the @Nullable: The prepareButton methods shouldnt return null and I changed the ScalingXXButtons to actually work correctly when passing a null image. I opted for a comment on passing null rather than the annotation as the comment is more helpful. I didnt fetch the changes I committed directly here (this always throws me) so they may show up again in the latest commit.

Windchild292 commented on megamek/src/megamek/client/ui/swing/BoardEditor.java:
```diff
@@ -299,9 +279,19 @@ public void setTerrainTypes(List<TerrainTypeHelper> terrainTypes) {
+    private FixedYPanel panelBoardSettings = new FixedYPanel();
+    
+    // Help Texts
+    private JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"),SwingConstants.LEFT);
+    private JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"),SwingConstants.LEFT);
```
```suggestion
    private JLabel labHelp1 = new JLabel(Messages.getString("BoardEditor.helpText"), SwingConstants.LEFT);
    private JLabel labHelp2 = new JLabel(Messages.getString("BoardEditor.helpText2"), SwingConstants.LEFT);
```

Windchild292 reviewed (commented):
One last piece, otherwise it looks good. I'd still like the nullable annotation just for a quick glance, but the comment is good :)
